File: packages/shared-types/src/ocr.ts
// Types for OCR extraction results from scanned instruction forms.
export type OCRResult = {
  ocr_result_id: string;
  source_image: string;
  confidence: number;
  fields: {
    initial_exam_room: string | null;
    ordered_services: string[];
    return_room: string | null;
    detected_room_codes: string[];
    room_descriptions: Record<string, string>;
    room_notes: Record<string, string>;
    room_queue_numbers: Record<string, string>;
  };
  requires_user_confirmation: boolean;
  // True when confidence is below threshold → UI shows the "chụp lại" fallback.
  is_low_confidence: boolean;
};

File: services/navigation-engine/app/models/ocr.py
"""OCR result models (mirror shared-types/ocr.ts).

The ``fields`` here are exactly what a confirmed instruction form contributes to a
session, so they line up 1:1 with ConfirmOcrFields used by /session/confirm-ocr.
"""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class OCRFields(BaseModel):
    initial_exam_room: Optional[str] = None
    ordered_services: list[str] = Field(default_factory=list)
    return_room: Optional[str] = None
    detected_room_codes: list[str] = Field(default_factory=list)
    room_descriptions: dict[str, str] = Field(default_factory=dict)
    room_notes: dict[str, str] = Field(default_factory=dict)
    room_queue_numbers: dict[str, str] = Field(default_factory=dict)


class OCRResult(BaseModel):
    ocr_result_id: str
    source_image: str
    confidence: float
    fields: OCRFields
    # The patient should always confirm the extracted fields before we use them.
    requires_user_confirmation: bool = True
    # True when confidence is below threshold → the UI shows the "chụp lại" fallback.
    is_low_confidence: bool = False
